File: src/main/irc/commands.ts
import type { IRCClient } from './client'
import { isupportNumber, fitsLimit } from '@shared/isupport'
import { parsePrefix, quietMode, banMask } from '@shared/powers'
import { maskToSet } from '@shared/masklists'

/**
 * Slash commands typed into the composer.
 *
 * Anything starting with a single "/" is a command. Unrecognised ones are
 * reported back to the user rather than sent to the channel — typing
 * "/msg NickServ IDENTIFY hunter2" must never end up as a public message.
 * A doubled slash ("//hello") escapes, and sends a literal "/hello".
 */

export interface CommandResult {
  /** False when the text is an ordinary message and should be sent as one */
  handled: boolean
  /** Text to send as a message instead (used by the // escape) */
  message?: string
  /** Set when the command could not be run */
  error?: string
}

const CHANNEL_PREFIXES = '#&+!'

const isChannel = (name: string): boolean => !!name && CHANNEL_PREFIXES.includes(name[0])

/**
 * Refuse text the server would silently cut, and say by how much.
 *
 * `TOPICLEN`, `AWAYLEN` and `KICKLEN` are not refusals — go over one and the
 * server accepts the command and quietly keeps the first N bytes. The user
 * finds out later, if at all. Better to say so now and let them decide what to
 * cut, since they are the only one who knows which half mattered.
 */
function tooLongFor(
  client: { state: { isupport: Record<string, string | true> } },
  token: string,
  /** The whole noun phrase, article and all — "an away message", not "away message" */
  what: string,
  text: string
): string | null {
  const limit = isupportNumber(client.state.isupport, token)
  if (fitsLimit(text, limit)) return null

  const length = Buffer.byteLength(text, 'utf8')
  return `This network allows ${limit} characters in ${what} and yours is ${length}.`
}

export function runCommand(client: IRCClient, target: string, text: string): CommandResult {
  if (!text.startsWith('/')) return { handled: false }

  // "//text" sends a literal message beginning with a slash
  if (text.startsWith('//')) return { handled: false, message: text.slice(1) }

  const spaceIdx = text.indexOf(' ')
  const name = (spaceIdx === -1 ? text.slice(1) : text.slice(1, spaceIdx)).toLowerCase()
  const rest = spaceIdx === -1 ? '' : text.slice(spaceIdx + 1).trim()
  const args = rest.length > 0 ? rest.split(/\s+/) : []

  /** Splits off the first word, keeping the remainder intact */
  const firstAndRest = (): [string, string] => {
    const idx = rest.indexOf(' ')
    return idx === -1 ? [rest, ''] : [rest.slice(0, idx), rest.slice(idx + 1).trim()]
  }

  switch (name) {
    /**
     * Send text as a message, whatever it starts with.
     *
     * What makes an alias able to produce a line beginning with a slash, and
     * the reason `//` exists as an escape in the first place.
     */
    case 'say': {
      if (!rest) return { handled: true, error: '/say needs something to say' }
      return { handled: false, message: rest }
    }

    case 'me': {
      if (!rest) return { handled: true, error: '/me needs something to do' }
      client.action(target, rest)
      return { handled: true }
    }

    case 'join':
    case 'j': {
      if (!args[0]) return { handled: true, error: 'Usage: /join #channel [key]' }
      const channel = isChannel(args[0]) ? args[0] : `#${args[0]}`
      client.join(channel, args[1])
      return { handled: true }
    }

    case 'part':
    case 'leave': {
      const channel = args[0] && isChannel(args[0]) ? args[0] : target
      const reason = args[0] && isChannel(args[0]) ? args.slice(1).join(' ') : rest
      if (!isChannel(channel)) return { handled: true, error: 'Usage: /part #channel [reason]' }
      client.part(channel, reason || undefined)
      return { handled: true }
    }

    case 'nick': {
      if (!args[0]) return { handled: true, error: 'Usage: /nick <nickname>' }
      client.setNick(args[0])
      return { handled: true }
    }

    case 'msg':
    case 'query': {
      const [to, body] = firstAndRest()
      if (!to) return { handled: true, error: 'Usage: /msg <nick|#channel> <message>' }
      if (!body) return { handled: true, error: `Nothing to send to ${to}` }
      client.say(to, body)
      return { handled: true }
    }

    case 'notice': {
      const [to, body] = firstAndRest()
      if (!to || !body) return { handled: true, error: 'Usage: /notice <target> <message>' }
      client.notice(to, body)
      return { handled: true }
    }

    case 'whois': {
      const nick = args[0]
      if (!nick) return { handled: true, error: 'Usage: /whois <nick>' }
      client.whois(nick)
      return { handled: true }
    }

    case 'topic': {
      if (!isChannel(target)) return { handled: true, error: '/topic only works in a channel' }
      if (!rest) {
        client.connection.send('TOPIC', target)
        return { handled: true }
      }
      const tooLong = tooLongFor(client, 'TOPICLEN', 'a topic', rest)
      if (tooLong) return { handled: true, error: tooLong }

      client.setTopic(target, rest)
      return { handled: true }
    }

    case 'mode': {
      if (args.length === 0) return { handled: true, error: 'Usage: /mode <target> <modes>' }

      // "/mode +o nick" applies to the current channel. Note a mode string and a
      // channel name can both start with "+", so match the mode shape first.
      if (/^[+-][a-zA-Z]+$/.test(args[0])) {
        client.mode(target, args[0], ...args.slice(1))
      } else if (args.length === 1) {
        client.connection.send('MODE', args[0])
      } else {
        client.mode(args[0], args[1], ...args.slice(2))
      }
      return { handled: true }
    }

    case 'kick': {
      if (!isChannel(target)) return { handled: true, error: '/kick only works in a channel' }
      const [nick, reason] = firstAndRest()
      if (!nick) return { handled: true, error: 'Usage: /kick <nick> [reason]' }

      const kickTooLong = reason && tooLongFor(client, 'KICKLEN', 'a kick reason', reason)
      if (kickTooLong) return { handled: true, error: kickTooLong }

      client.kick(target, nick, reason || undefined)
      return { handled: true }
    }

    case 'invite': {
      const [nick, channel] = firstAndRest()
      if (!nick) return { handled: true, error: 'Usage: /invite <nick> [#channel]' }
      client.connection.send('INVITE', nick, channel || target)
      return { handled: true }
    }

    case 'away': {
      // AWAY with no message clears it
      if (!rest) {
        client.connection.send('AWAY')
        return { handled: true }
      }

      const awayTooLong = tooLongFor(client, 'AWAYLEN', 'an away message', rest)
      if (awayTooLong) return { handled: true, error: awayTooLong }

      client.connection.send('AWAY', rest)
      return { handled: true }
    }

    case 'back': {
      client.connection.send('AWAY')
      return { handled: true }
    }

    case 'quit': {
      client.disconnect(rest || undefined)
      return { handled: true }
    }

    /**
     * Become a server operator.
     *
     * The password goes on the wire and nowhere else: the debug stream
     * redacts it (`@shared/redact`) and that stream does not leave this
     * machine at all. Worth saying because this is the one command here whose
     * argument is a credential.
     */
    case 'oper': {
      const [name, ...password] = rest.split(' ')
      if (!name || password.length === 0) {
        return { handled: true, error: 'Usage: /oper <name> <password>' }
      }
      client.connection.send('OPER', name, password.join(' '))
      return { handled: true }
    }

    case 'raw':
    case 'quote': {
      if (!rest) return { handled: true, error: 'Usage: /raw <IRC line>' }
      client.connection.sendRaw(rest)
      return { handled: true }
    }

    // ── Ranks ──────────────────────────────────────────────────────
    //
    // The everyday ones. Every client has had these for thirty years and this
    // one made you type `/mode #channel +o nick`, which is the same thing with
    // more to get wrong.

    case 'op':
    case 'deop':
    case 'voice':
    case 'devoice':
    case 'halfop':
    case 'dehalfop':
    case 'owner':
    case 'deowner':
    case 'admin':
    case 'deadmin': {
      const letters: Record<string, string> = {
        op: 'o',
        voice: 'v',
        halfop: 'h',
        owner: 'q',
        admin: 'a'
      }
      const adding = !name.startsWith('de')
      const letter = letters[adding ? name : name.slice(2)]
      const channel = isChannel(args[0]) ? args.shift()! : target
      if (!isChannel(channel)) return { handled: true, error: `/${name} only works in a channel` }

      const nicks = args.length > 0 ? args : [client.state.nick]
      // One MODE per MODES-worth, because a server that takes four at a time
      // silently drops the fifth.
      const perLine = isupportNumber(client.state.isupport, 'MODES') ?? 4
      for (let at = 0; at < nicks.length; at += perLine) {
        const batch = nicks.slice(at, at + perLine)
        client.connection.send(
          'MODE',
          channel,
          `${adding ? '+' : '-'}${letter.repeat(batch.length)}`,
          ...batch
        )
      }
      return { handled: true }
    }

    // ── Lists ──────────────────────────────────────────────────────

    case 'ban':
    case 'unban':
    case 'quiet':
    case 'unquiet': {
      const scheme = parsePrefix(client.state.isupport.PREFIX as string)
      const quieting = name.endsWith('quiet')
      const letter = quieting ? quietMode(client.state.isupport.CHANMODES as string, scheme) : 'b'
      if (!letter) {
        return { handled: true, error: 'This network does not have a quiet mode' }
      }

      const channel = isChannel(args[0]) ? args.shift()! : target
      if (!isChannel(channel)) return { handled: true, error: `/${name} only works in a channel` }
      if (args.length === 0) return { handled: true, error: `Usage: /${name} <nick or mask>` }

      const adding = !name.startsWith('un')
      for (const who of args) {
        // A bare nick becomes `nick!*@*`, which is what almost every server
        // would have done anyway and what the person plainly meant.
        client.connection.send('MODE', channel, `${adding ? '+' : '-'}${letter}`, maskToSet(who))
      }
      return { handled: true }
    }

    case 'kickban': {
      const channel = isChannel(args[0]) ? args.shift()! : target
      if (!isChannel(channel)) return { handled: true, error: '/kickban only works in a channel' }
      const who = args.shift()
      if (!who) return { handled: true, error: 'Usage: /kickban <nick> [reason]' }

      // Ban first. Kicking first leaves a window — short, but real — in which
      // they can rejoin before the ban lands, which is the one thing kickban
      // exists to prevent.
      const user = client.state.channels
        .get(client.state.casemap(channel))
        ?.users.get(client.state.casemap(who))
      client.connection.send('MODE', channel, '+b', banMask({ nick: who, host: user?.host }))
      client.connection.send('KICK', channel, who, args.join(' ') || who)
      return { handled: true }
    }

    case 'banlist':
    case 'bans': {
      const channel = isChannel(args[0]) ? args[0] : target
      if (!isChannel(channel)) return { handled: true, error: '/banlist only works in a channel' }
      client.connection.send('MODE', channel, '+b')
      return { handled: true }
    }

    // ── Channels ───────────────────────────────────────────────────

    case 'cycle':
    case 'hop': {
      const channel = isChannel(args[0]) ? args[0] : target
      if (!isChannel(channel)) return { handled: true, error: `/${name} only works in a channel` }
      client.connection.send('PART', channel)
      client.connection.send('JOIN', channel)
      return { handled: true }
    }

    case 'knock': {
      const [channel, ...why] = args
      if (!isChannel(channel)) return { handled: true, error: 'Usage: /knock <#channel> [reason]' }
      client.connection.send('KNOCK', channel, why.join(' ') || 'Please let me in')
      return { handled: true }
    }

    case 'names': {
      const channel = isChannel(args[0]) ? args[0] : target
      if (!isChannel(channel)) return { handled: true, error: '/names only works in a channel' }
      client.connection.send('NAMES', channel)
      return { handled: true }
    }

    case 'list': {
      // With no arguments this is every channel on the network, which on a
      // large one is tens of thousands of lines. The browser asks the same
      // question with a filter box, so send them there.
      if (args.length > 0) client.connection.send('LIST', ...args)
      else client.connection.send('LIST')
      return { handled: true }
    }

    // ── People ─────────────────────────────────────────────────────

    case 'whowas': {
      if (!rest) return { handled: true, error: 'Usage: /whowas <nick> [count]' }
      client.connection.send('WHOWAS', ...args)
      return { handled: true }
    }

    case 'who': {
      if (!rest) return { handled: true, error: 'Usage: /who <nick, #channel or mask>' }
      client.connection.send('WHO', ...args)
      return { handled: true }
    }

    case 'ctcp': {
      const [who, verb, ...body] = args
      if (!who || !verb) return { handled: true, error: 'Usage: /ctcp <nick> <VERSION|PING|TIME|…>' }
      client.connection.sendRaw(
        `PRIVMSG ${who} :\u0001${verb.toUpperCase()}${body.length ? ' ' + body.join(' ') : ''}\u0001`
      )
      return { handled: true }
    }

    case 'setname': {
      if (!rest) return { handled: true, error: 'Usage: /setname <real name>' }
      client.connection.send('SETNAME', rest)
      return { handled: true }
    }

    // ── The server ─────────────────────────────────────────────────

    case 'motd':
    case 'lusers':
    case 'time':
    case 'version':
    case 'links':
    case 'stats':
    case 'info':
    case 'admininfo': {
      const verb = name === 'admininfo' ? 'ADMIN' : name.toUpperCase()
      if (args.length > 0) client.connection.send(verb, ...args)
      else client.connection.send(verb)
      return { handled: true }
    }

    case 'wallops': {
      if (!rest) return { handled: true, error: 'Usage: /wallops <message>' }
      client.connection.send('WALLOPS', rest)
      return { handled: true }
    }

    case 'ping': {
      const who = args[0]
      if (who) {
        // A CTCP PING to a person, which is what /ping means everywhere —
        // stamped so the reply can be turned back into a round trip.
        client.connection.sendRaw(`PRIVMSG ${who} :\u0001PING ${Date.now()}\u0001`)
      } else {
        client.connection.send('PING', String(Date.now()))
      }
      return { handled: true }
    }

    default:
      return { handled: true, error: `Unknown command: /${name}` }
  }
}

File: android/app/src/main/java/org/switchboard/android/irc/Commands.kt
package org.switchboard.android.irc

/**
 * Slash commands typed into the composer.
 *
 * The phone had none. Following a desktop that did not matter — `message:send`
 * went over the link and `commands.ts` ran there — but holding the connection
 * itself, every one of them was sent to the channel as ordinary text. Typing
 * `/msg NickServ IDENTIFY hunter2` published the password to the room.
 *
 * So this mirrors `src/main/irc/commands.ts`, command for command and message
 * for message: the same client on two devices should not answer `/topic`
 * differently depending on which one you picked up.
 */
/**
 * What a slash command is allowed to do.
 *
 * The same seam as [IrcSession], one level up: that one is what a *protocol
 * handler* may do, this is what a *typed command* may do. [IrcConnection]
 * implements it, and a test can stand in for it without a socket — which
 * matters here more than most places, because the bug this file exists to fix
 * was a password going to a channel.
 */
interface IrcCommandTarget {
    val state: ConnectionState

    fun send(command: String, vararg params: String)
    fun sendRaw(line: String)
    fun say(target: String, text: String)
    fun action(target: String, text: String)
    fun notice(target: String, text: String)
    fun join(channel: String, key: String? = null)
    fun part(channel: String, reason: String? = null)
    fun setTopic(channel: String, topic: String)
    fun setNick(nick: String)
    fun whois(nick: String)
    fun setAway(message: String?)
    fun setMode(target: String, mode: String, vararg args: String)
    fun kick(channel: String, nick: String, reason: String?)
    fun invite(nick: String, channel: String)
    fun stop(quitMessage: String = "Switchboard")
}

object Commands {

    /**
     * What to do with what was typed.
     *
     * [handled] false means it was never a command and should be sent as a
     * message; [message] is the text to send instead, which the `//` escape
     * uses; [error] is something to tell the user rather than the network.
     */
    data class Result(
        val handled: Boolean,
        val message: String? = null,
        val error: String? = null
    )

    private const val CHANNEL_PREFIXES = "#&+!"

    private fun isChannel(name: String): Boolean =
        name.isNotEmpty() && CHANNEL_PREFIXES.contains(name[0])

    /**
     * Refuse text the server would silently cut, and say by how much.
     *
     * `TOPICLEN`, `AWAYLEN` and `KICKLEN` are not refusals — go over one and
     * the server accepts the command and quietly keeps the first N bytes. The
     * user finds out later, if at all.
     */
    private fun tooLongFor(
        connection: IrcCommandTarget,
        token: String,
        /** The whole noun phrase, article and all — "an away message" */
        what: String,
        text: String
    ): String? {
        val limit = Isupport.number(connection.state.isupport, token)
        if (Isupport.fits(text, limit)) return null

        val length = text.toByteArray(Charsets.UTF_8).size
        return "This network allows $limit characters in $what and yours is $length."
    }

    fun run(connection: IrcCommandTarget, target: String, text: String): Result {
        if (!text.startsWith("/")) return Result(handled = false)

        // "//text" sends a literal message beginning with a slash
        if (text.startsWith("//")) return Result(handled = false, message = text.substring(1))

        val space = text.indexOf(' ')
        val name = (if (space == -1) text.substring(1) else text.substring(1, space)).lowercase()
        val rest = if (space == -1) "" else text.substring(space + 1).trim()
        val args = if (rest.isEmpty()) emptyList() else rest.split(Regex("\\s+"))

        /** Splits off the first word, keeping the remainder intact */
        fun firstAndRest(): Pair<String, String> {
            val at = rest.indexOf(' ')
            return if (at == -1) rest to ""
            else rest.substring(0, at) to rest.substring(at + 1).trim()
        }

        return when (name) {
            "me" -> {
                if (rest.isEmpty()) Result(true, error = "/me needs something to do")
                else {
                    connection.action(target, rest)
                    Result(true)
                }
            }

            "join", "j" -> {
                val first = args.getOrNull(0)
                if (first == null) Result(true, error = "Usage: /join #channel [key]")
                else {
                    connection.join(if (isChannel(first)) first else "#$first", args.getOrNull(1))
                    Result(true)
                }
            }

            "part", "leave" -> {
                val named = args.getOrNull(0)?.takeIf { isChannel(it) }
                val channel = named ?: target
                val reason = if (named != null) args.drop(1).joinToString(" ") else rest
                if (!isChannel(channel)) Result(true, error = "Usage: /part #channel [reason]")
                else {
                    connection.part(channel, reason.ifBlank { null })
                    Result(true)
                }
            }

            "nick" -> {
                val nick = args.getOrNull(0)
                if (nick == null) Result(true, error = "Usage: /nick <nickname>")
                else {
                    connection.setNick(nick)
                    Result(true)
                }
            }

            "msg", "query" -> {
                val (to, body) = firstAndRest()
                when {
                    to.isEmpty() -> Result(true, error = "Usage: /msg <nick|#channel> <message>")
                    body.isEmpty() -> Result(true, error = "Nothing to send to $to")
                    else -> {
                        connection.say(to, body)
                        Result(true)
                    }
                }
            }

            "notice" -> {
                val (to, body) = firstAndRest()
                if (to.isEmpty() || body.isEmpty()) {
                    Result(true, error = "Usage: /notice <target> <message>")
                } else {
                    connection.notice(to, body)
                    Result(true)
                }
            }

            "whois" -> {
                val nick = args.getOrNull(0)
                if (nick == null) Result(true, error = "Usage: /whois <nick>")
                else {
                    connection.whois(nick)
                    Result(true)
                }
            }

            "topic" -> {
                if (!isChannel(target)) Result(true, error = "/topic only works in a channel")
                else if (rest.isEmpty()) {
                    connection.send("TOPIC", target)
                    Result(true)
                } else {
                    val tooLong = tooLongFor(connection, "TOPICLEN", "a topic", rest)
                    if (tooLong != null) Result(true, error = tooLong)
                    else {
                        connection.setTopic(target, rest)
                        Result(true)
                    }
                }
            }

            "mode" -> {
                if (args.isEmpty()) Result(true, error = "Usage: /mode <target> <modes>")
                else {
                    // "/mode +o nick" applies to the current channel. A mode
                    // string and a channel name can both start with "+", so
                    // match the mode shape first.
                    if (Regex("^[+-][a-zA-Z]+$").matches(args[0])) {
                        connection.setMode(target, args[0], *args.drop(1).toTypedArray())
                    } else if (args.size == 1) {
                        connection.send("MODE", args[0])
                    } else {
                        connection.setMode(args[0], args[1], *args.drop(2).toTypedArray())
                    }
                    Result(true)
                }
            }

            "kick" -> {
                if (!isChannel(target)) Result(true, error = "/kick only works in a channel")
                else {
                    val (nick, reason) = firstAndRest()
                    val tooLong = reason.takeIf { it.isNotEmpty() }
                        ?.let { tooLongFor(connection, "KICKLEN", "a kick reason", it) }
                    when {
                        nick.isEmpty() -> Result(true, error = "Usage: /kick <nick> [reason]")
                        tooLong != null -> Result(true, error = tooLong)
                        else -> {
                            connection.kick(target, nick, reason.ifBlank { null })
                            Result(true)
                        }
                    }
                }
            }

            "invite" -> {
                val (nick, channel) = firstAndRest()
                if (nick.isEmpty()) Result(true, error = "Usage: /invite <nick> [#channel]")
                else {
                    connection.invite(nick, channel.ifBlank { target })
                    Result(true)
                }
            }

            "away" -> {
                // AWAY with no message clears it
                if (rest.isEmpty()) {
                    connection.setAway(null)
                    Result(true)
                } else {
                    val tooLong = tooLongFor(connection, "AWAYLEN", "an away message", rest)
                    if (tooLong != null) Result(true, error = tooLong)
                    else {
                        connection.setAway(rest)
                        Result(true)
                    }
                }
            }

            "back" -> {
                connection.setAway(null)
                Result(true)
            }

            "quit" -> {
                connection.stop(rest.ifBlank { "Switchboard" })
                Result(true)
            }

            /**
             * Become a server operator.
             *
             * The password goes on the wire and nowhere else — the one
             * command here whose argument is a credential.
             */
            "oper" -> {
                val name = rest.substringBefore(' ').trim()
                val password = rest.substringAfter(' ', "").trim()
                if (name.isEmpty() || password.isEmpty()) {
                    Result(true, error = "Usage: /oper <name> <password>")
                } else {
                    connection.send("OPER", name, password)
                    Result(true)
                }
            }

            /**
             * Send text as a message, whatever it starts with.
             *
             * What makes an alias able to produce a line beginning with a
             * slash, and the reason `//` exists as an escape in the first
             * place.
             */
            "say" -> {
                if (rest.isEmpty()) Result(true, error = "/say needs something to say")
                else Result(false, message = rest)
            }

            "raw", "quote" -> {
                if (rest.isEmpty()) Result(true, error = "Usage: /raw <IRC line>")
                else {
                    connection.sendRaw(rest)
                    Result(true)
                }
            }

            // ── Ranks ──────────────────────────────────────────────
            //
            // The everyday ones. Every client has had these for thirty years
            // and this one made you type `/mode #channel +o nick`, which is
            // the same thing with more to get wrong.
            "op", "deop", "voice", "devoice", "halfop", "dehalfop",
            "owner", "deowner", "admin", "deadmin" -> {
                val letters = mapOf("op" to "o", "voice" to "v", "halfop" to "h",
                    "owner" to "q", "admin" to "a")
                val adding = !name.startsWith("de")
                val letter = letters[if (adding) name else name.drop(2)]!!
                val words = args.toMutableList()
                val channel = if (isChannel(words.firstOrNull().orEmpty())) words.removeAt(0) else target

                if (!isChannel(channel)) Result(true, error = "/$name only works in a channel")
                else {
                    val nicks = words.ifEmpty { listOf(connection.state.nick) }
                    // One MODE per MODES-worth: a server that takes four at a
                    // time silently drops the fifth.
                    val perLine = Isupport.number(connection.state.isupport, "MODES") ?: 4
                    nicks.chunked(perLine).forEach { batch ->
                        connection.setMode(
                            channel,
                            (if (adding) "+" else "-") + letter.repeat(batch.size),
                            *batch.toTypedArray()
                        )
                    }
                    Result(true)
                }
            }

            // ── Lists ──────────────────────────────────────────────
            "ban", "unban", "quiet", "unquiet" -> {
                val scheme = Powers.parsePrefix(connection.state.isupport["PREFIX"])
                val quieting = name.endsWith("quiet")
                val letter: String? = if (quieting)
                    Powers.quietMode(connection.state.isupport["CHANMODES"], scheme)?.toString()
                else "b"

                val words = args.toMutableList()
                val channel = if (isChannel(words.firstOrNull().orEmpty())) words.removeAt(0) else target

                when {
                    letter == null -> Result(true, error = "This network does not have a quiet mode")
                    !isChannel(channel) -> Result(true, error = "/$name only works in a channel")
                    words.isEmpty() -> Result(true, error = "Usage: /$name <nick or mask>")
                    else -> {
                        val adding = !name.startsWith("un")
                        // A bare nick becomes `nick!*@*`, which is what almost
                        // every server would have done anyway.
                        words.forEach { who ->
                            connection.setMode(
                                channel,
                                (if (adding) "+" else "-") + letter,
                                MaskLists.maskToSet(who)
                            )
                        }
                        Result(true)
                    }
                }
            }

            "kickban" -> {
                val words = args.toMutableList()
                val channel = if (isChannel(words.firstOrNull().orEmpty())) words.removeAt(0) else target
                val who = if (words.isEmpty()) null else words.removeAt(0)

                when {
                    !isChannel(channel) -> Result(true, error = "/kickban only works in a channel")
                    who == null -> Result(true, error = "Usage: /kickban <nick> [reason]")
                    else -> {
                        // Ban first. Kicking first leaves a window — short but
                        // real — in which they can rejoin before the ban lands,
                        // which is the one thing kickban exists to prevent.
                        val host = connection.state.findChannel(channel)
                            ?.users?.get(connection.state.casemap(who))?.host
                        connection.setMode(channel, "+b", Powers.banMask(who, host))
                        connection.kick(channel, who, words.joinToString(" ").ifEmpty { who })
                        Result(true)
                    }
                }
            }

            "banlist", "bans" -> {
                val channel = if (isChannel(args.firstOrNull().orEmpty())) args[0] else target
                if (!isChannel(channel)) Result(true, error = "/banlist only works in a channel")
                else {
                    connection.setMode(channel, "+b")
                    Result(true)
                }
            }

            // ── Channels ───────────────────────────────────────────
            "cycle", "hop" -> {
                val channel = if (isChannel(args.firstOrNull().orEmpty())) args[0] else target
                if (!isChannel(channel)) Result(true, error = "/$name only works in a channel")
                else {
                    connection.part(channel)
                    connection.join(channel)
                    Result(true)
                }
            }

            "knock" -> {
                val channel = args.firstOrNull().orEmpty()
                if (!isChannel(channel)) Result(true, error = "Usage: /knock <#channel> [reason]")
                else {
                    val why = args.drop(1).joinToString(" ").ifEmpty { "Please let me in" }
                    connection.send("KNOCK", channel, why)
                    Result(true)
                }
            }

            "names" -> {
                val channel = if (isChannel(args.firstOrNull().orEmpty())) args[0] else target
                if (!isChannel(channel)) Result(true, error = "/names only works in a channel")
                else {
                    connection.send("NAMES", channel)
                    Result(true)
                }
            }

            "list" -> {
                // With no arguments this is every channel on the network,
                // which on a large one is tens of thousands of lines.
                connection.send("LIST", *args.toTypedArray())
                Result(true)
            }

            // ── People ─────────────────────────────────────────────
            "whowas" -> {
                if (rest.isEmpty()) Result(true, error = "Usage: /whowas <nick> [count]")
                else {
                    connection.send("WHOWAS", *args.toTypedArray())
                    Result(true)
                }
            }

            "who" -> {
                if (rest.isEmpty()) Result(true, error = "Usage: /who <nick, #channel or mask>")
                else {
                    connection.send("WHO", *args.toTypedArray())
                    Result(true)
                }
            }

            "ctcp" -> {
                val who = args.getOrNull(0)
                val verb = args.getOrNull(1)
                if (who == null || verb == null)
                    Result(true, error = "Usage: /ctcp <nick> <VERSION|PING|TIME|…>")
                else {
                    val body = args.drop(2).joinToString(" ")
                    val payload = verb.uppercase() + if (body.isEmpty()) "" else " $body"
                    connection.sendRaw("PRIVMSG $who :\u0001$payload\u0001")
                    Result(true)
                }
            }

            "setname" -> {
                if (rest.isEmpty()) Result(true, error = "Usage: /setname <real name>")
                else {
                    connection.send("SETNAME", rest)
                    Result(true)
                }
            }

            // ── The server ─────────────────────────────────────────
            "motd", "lusers", "time", "version", "links", "stats", "info", "admininfo" -> {
                val verb = if (name == "admininfo") "ADMIN" else name.uppercase()
                connection.send(verb, *args.toTypedArray())
                Result(true)
            }

            "wallops" -> {
                if (rest.isEmpty()) Result(true, error = "Usage: /wallops <message>")
                else {
                    connection.send("WALLOPS", rest)
                    Result(true)
                }
            }

            "ping" -> {
                val who = args.firstOrNull()
                if (who != null) {
                    // A CTCP PING to a person, which is what /ping means
                    // everywhere — stamped so the reply is a round trip.
                    connection.sendRaw("PRIVMSG $who :\u0001PING ${System.currentTimeMillis()}\u0001")
                } else {
                    connection.send("PING", System.currentTimeMillis().toString())
                }
                Result(true)
            }

            else -> Result(true, error = "Unknown command: /$name")
        }
    }
}
